File: tictactoe/connectfour.py
__author__ = 'Mark'

import random

class ConnectFourGameState(object):

    def __init__(self, width=7, height=6):
        self.currentPlayer = 1
        self.board = [[0] * width for _ in xrange(height)]
        self.width = width
        self.height = height
        self.winner = None

    def copy(self):
        gamestate = ConnectFourGameState()
        gamestate.board = [list(row) for row in self.board]
        gamestate.currentPlayer = self.currentPlayer
        gamestate.winner = self.winner

        return gamestate

    def executeMove(self, move):
        assert self.winner is None
        column = [row[move] for row in self.board]

        r = self.height - 1

        for i, c in enumerate(column):
            if c != 0:
                r = i - 1
                break

        self.board[r][move] = self.currentPlayer
        self.checkForWin(r, move, self.currentPlayer)
        self.currentPlayer = 3 - self.currentPlayer


    def getLegalMoves(self):

        first_row = self.board[0]
        legalMoves = [i for i, v in enumerate(first_row) if v == 0]

        return legalMoves


    def checkForWin(self, _r, _c, player):

        rmin = self._clamp(_r - 3, 0, self.height)
        rmax = self._clamp(_r + 3, 0, self.height - 1)
        cmin = self._clamp(_c - 3, 0, self.width)
        cmax = self._clamp(_c + 3, 0, self.width - 1)
        # Check only cells adjacent to the last played move

        row = self.board[_r]
        col = [_row[_c] for _row in self.board]

        count = 0
        for c in xrange(cmin, cmax + 1):
            if row[c] == player:
                count += 1
            else:
                count = 0

            if count == 4:
                self.winner = player

        count = 0
        for r in xrange(rmin, rmax + 1):
            if col[r] == player:
                count += 1
            else:
                count = 0

            if count == 4:
                self.winner = player

        count = 0
        for r, c in zip(xrange(rmax, rmin - 1, -1), xrange(cmin, cmax + 1)):
            if self.board[r][c] == player:
                count += 1
            else:
                count = 0

            if count == 4:
                self.winner = player

        count = 0
        for r, c in zip(xrange(rmin, rmax + 1), xrange(cmin, cmax + 1)):
            if self.board[r][c] == player:
                count += 1
            else:
                count = 0

            if count == 4:
                self.winner = player

        # If the top row has no empty slots, its a draw
        if 0 not in self.board[0]:
            self.winner = 0


    def _clamp(self, value, minimum, maximum):
        return max(minimum, min(value, maximum))


    def __repr__(self):
        out = []
        icons = ['-', 'x', 'o']
        iconBoard = [[icons[c] for c in row] for row in self.board]

        for row in iconBoard:
            row.append("\n")
            out.append(''.join(row))

        return ''.join(out)


def playRandomMoves(num_moves = 10):
    moves = [random.randint(0, 6) for _ in xrange(num_moves)]

    for move in moves:
        cf.executeMove(move)

if __name__ == "__main__":
    cf = ConnectFourGameState()
    #playRandomMoves()
    cf.executeMove(3)
    cf.executeMove(4)
    cf.executeMove(4)
    cf.executeMove(5)
    cf.executeMove(5)
    cf.executeMove(6)
    cf.executeMove(5)
    cf.executeMove(6)
    cf.executeMove(6)
    cf.executeMove(0)
    cf.executeMove(6)


    print cf
    print cf.getLegalMoves()

File: tictactoe/web/connectfour.js
var WIDTH = 7;
var HEIGHT = 6;

function Gamestate() {
  //TODO: Make this dynamically created
  this.board = [
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0],
  ];

  this.currentPlayer = 1;
  this.width = WIDTH;
  this.height = HEIGHT;
  this.winner = null;

}

Gamestate.prototype.executeMove = function(move) {
  var column = [];
  var row = -1;

  for (var r = 0; r < this.board.length; r++) {
    column.push(this.board[r][move]);
  }

  for (var r = column.length - 1; r >= 0; r--) {
    if (column[r] == 0) {
      row = r;
      break;
    }
  }

  if (row != -1) {
    this.board[row][move] = this.currentPlayer;
    this.checkForWin(row, move, this.currentPlayer);
    this.currentPlayer = 3 - this.currentPlayer;
  }

  return row;

}

Gamestate.prototype.getLegalMoves = function() {

  var first_row = this.board[0];
  var legalMoves = [];

  for (var i = 0; i < first_row.length; i++) {
    if (first_row[i] == 0) {
      legalMoves.push(i);
    }
  }

  return legalMoves;

}

Gamestate.prototype.checkForWin = function(_r, _c, player) {

  var rmin = clamp(_r - 3, 0, this.height);
  var rmax = clamp(_r + 3, 0, this.height - 1);
  var cmin = clamp(_c - 3, 0, this.width);
  var cmax = clamp(_c + 3, 0, this.width - 1);

  row = this.board[_r];
  var col = [];
  for (var r = 0; r < this.board.length; r++) {
    col.push(this.board[r][_c]);
  }

  // Check column for victory
  var count = 0;
  for (var c = cmin; c <= cmax; c++) {
    if (row[c] == player) {
      count += 1;
    } else {
      count = 0;
    }

    if (count == 4) {
      this.winner = player;
    }
  }

  // Check row for victory
  var count = 0;
  for (var r = rmin; r <= rmax; r++) {
    if (col[r] == player) {
      count += 1;
    } else {
      count = 0;
    }

    if (count == 4) {
      this.winner = player;
    }
  }

  // Check diagonal direction -> /
  var count = 0;
  var row_indicies = [];
  var col_indicies = [];

  for (var i = rmax; i >= rmin; i--) {
    row_indicies.push(i);
  }

  for (var i = cmin; i <= cmax; i++) {
    col_indicies.push(i);
  }

  var indicies = zip(row_indicies, col_indicies);

  for (var i = 0; i < indicies.length; i++) {
    var r = indicies[i][0];
    var c = indicies[i][1];

    if (this.board[r][c] == player) {
      count += 1;
    } else {
      count = 0;
    }

    if (count == 4) {
      this.winner = player;
    }

  }

  // Check diagonal direction -> \
  var count = 0;
  var row_indicies = [];
  var col_indicies = [];

  for (var i = rmin; i <= rmax; i++) {
    row_indicies.push(i);
  }

  for (var i = cmin; i <= cmax; i++) {
    col_indicies.push(i);
  }

  var indicies = zip(row_indicies, col_indicies);

  for (var i = 0; i < indicies.length; i++) {
    var r = indicies[i][0];
    var c = indicies[i][1];

    if (this.board[r][c] == player) {
      count += 1;
    } else {
      count = 0;
    }

    if (count == 4) {
      this.winner = player;
    }

  }

  // If the top row has no empty slots and we got here, its a draw
  if (this.board[0].indexOf(0) == -1) {
    this.winner = 0;
  }


}

Gamestate.prototype.copy = function() {

  var gameState = new Gamestate();
  for (var r = 0; r < this.board.length; r++) {
    gameState.board[r] = this.board[r].slice(0);
  }
  gameState.currentPlayer = this.currentPlayer;
  gameState.winner = this.winner;

  return gameState;

}
